### PR TITLE
Add RTC (Real Time Clock) driver

### PR DESCRIPTION
**Pull request by PrayagS**
This PR proposes the device driver for RTC chips. 

The required message type (subject to change) has been added to our existing message interface. I agree with your proposal to use unixtime as the datatype. I'll be soon getting back to you with such an implementation.

Please let me know if there's a need to write a test application for this driver.

Thanks again for your continued support and prompt responses. I'm sure our whole team appreciates that.

### CODE REVIEW

File: servers/drivers/time/rtc/main.c
Comment (nuta) [inline servers/drivers/time/rtc/main.c:26]: How about defining a struct for its return value?

```c
struct datetime {
    uint8_t year;
    uint8_t month;
    ...
};

void read_datetime(struct datetime *datetime) {
```

File: servers/drivers/time/rtc/main.c
Comment (nuta) [inline servers/drivers/time/rtc/main.c:13]: Nit: Please avoid using uppercase characters for a variable. Use only lowercase ones. This line should be:

```c
    uint8_t status = io_read8(rtc_io, 1);
```

File: servers/drivers/time/rtc/main.c
Comment (nuta) [inline servers/drivers/time/rtc/main.c:11]: Nit: I think this function should return a boolean value.

File: servers/drivers/time/rtc/main.c
Comment (nuta) [inline servers/drivers/time/rtc/main.c:40]: Nit: Use lowercase characters for variables. This should be `use_bcd`.

File: tools/genkconfig.py
Comment (nuta) [inline tools/genkconfig.py:104]: How about `servers/drivers/rtc`?

File: interface.idl
Comment (nuta) [inline interface.idl:59]: `time` is a bit confusing with `datetime` server which we'll work on later. How about `namespace rtc`?

File: interface.idl
Comment (nuta) [inline interface.idl:60]: Returning string requires some allocations and is not efficient. Please return as integers:

```
    rpc read() -> (year: uint32, month: uint8, day: uint8, hour: uint8, minute: uint8, second: uint8, day_of_week: uint8);
```

File: tools/genkconfig.py
Comment (nuta) [inline tools/genkconfig.py:104]: Nit: `Timer Driver` -> `Real Time Clock Driers`

A *timer* sounds similar to RTC, but they're quite different.

**Review by nuta — CHANGES_REQUESTED** [review]
Thank you for the PR, @PrayagS! Could you consider my review comments?

File: servers/drivers/time/rtc/main.c
Comment (PrayagS) [reply servers/drivers/time/rtc/main.c:13]: Noted

File: servers/drivers/time/rtc/main.c
Comment (PrayagS) [reply servers/drivers/time/rtc/main.c:26]: Yeah, this is something I was debating. Should I define the struct for use in this driver only? Or should I define it as a global type in `types.h`? That would also allow using the same type in message passing. And it will be also useful when converting time from UNIX-like timestamp.

File: tools/genkconfig.py
Comment (PrayagS) [reply tools/genkconfig.py:104]: Makes sense. Changed.

File: servers/drivers/time/rtc/main.c
Comment (nuta) [reply servers/drivers/time/rtc/main.c:26]: Please define it for this driver only for now. Let's consider moving it into the global scope when we actually need to do. 

JFYI, the IPC stub generator does not yet support `struct` types in the message payload.

Comment (PrayagS) [reply servers/drivers/time/rtc/main.c:26]: Yeah, I did notice that the generator didn't support that. That's why I didn't go with struct. What you proposed seems the best for now.

**Comment by PrayagS** [comment]
I've modified the code as per the changes you requested. Please let me know if there's still some issue.

I had a couple of questions,
- The function `io_alloc_port` in `io.c`, has this argument `len`. Looking at the code, I'm unsure about how is it affecting the allocation.
  https://github.com/nuta/resea/blob/aff0d65e34c6597e62b9fc9ff9fac286f89dc8a7/libs/driver/io.c#L6-L11
  In the rtc driver, I've used a placeholder value for that.
  ```c
  rtc_io = io_alloc_port(RTC_PORT_IDX, 0x10, IO_ALLOC_NORMAL);
  ```
  I'm not sure what the correct value should be.
- What do you mean when you comment "Nit:"?

**Comment by nuta** [comment]
> The function io_alloc_port in io.c, has this argument len. Looking at the code, I'm unsure about how is it affecting the allocation.

It is used to limit the port/memory address range accessible from the `io_t` object. For this RTC device, it has two IO ports (0x70, 0x71) so `len` should be 2. Currently, as you can see, it is not used in the io-mapped IO (`IO_SPACE_IO`) since Resea allows accessing all IO ports by default.

> What do you mean when you comment "Nit:"?

It stands for "nitpick". I (and folks in the OSS community) use this word in code reviews to say *"I do know this is a trivial (and boring) change but I would be very grateful if you could fix it by yourself"*.

File: tools/genkconfig.py
Comment (nuta) [inline tools/genkconfig.py:7]: Nit: Do we need this change?

File: tools/genkconfig.py
Comment (nuta) [inline tools/genkconfig.py:100]: I think you don't need to add "servers/drivers" entry here.

File: tools/genkconfig.py
Comment (nuta) [inline tools/genkconfig.py:105]: Please move `servers/drivers/rtc` into `servers/drivers/rtc/rtc`.

**Review by nuta — CHANGES_REQUESTED** [review]
LGTM with some nitpicks.

File: tools/genkconfig.py
Comment (PrayagS) [reply tools/genkconfig.py:7]: Oh my bad. I forgot to save the file without formatting.

File: tools/genkconfig.py
Comment (PrayagS) [reply tools/genkconfig.py:100]: I did this to avoid the change you requested below. But now as you said, I'll move the files to `servers/drivers/rtc/rtc`.

**Comment by PrayagS** [comment]
Thanks for clearing out those questions. Please have a look again regarding the changes you requested.

File: servers/drivers/rtc/rtc/main.c
Comment (nuta) [inline servers/drivers/rtc/rtc/main.c:83]: Nit: Please remove this unnecessary code.

**Review by nuta — CHANGES_REQUESTED** [review]
I've tested this RTC driver on my environment and it works perfectly! But we need to fix a nit before merging...

File: servers/drivers/rtc/rtc/main.c
Comment (PrayagS) [reply servers/drivers/rtc/rtc/main.c]: Done.

**Review by nuta — APPROVED** [review]
LGTM. I'll merge this PR once CI tests pass.

**Comment by nuta** [comment]
Merged. Thank you for your work, @PrayagS 👍 👍

**Comment by PrayagS** [comment]
Thank you too for your continued support. I'll get back soon with progress on the `datetime` server.

**Pull request merged**
